tests/unit/s2n_tls13_handshake_test.c

```diff
@@ -0,0 +1,313 @@
+#include "tls/s2n_handshake_io.c"
+
+static message_type_t invalid_handshake[S2N_MAX_HANDSHAKE_LENGTH];
+
+static int s2n_test_handler(struct s2n_connection* conn)
+{
+    return 0;
+}
+
+int s2n_test_write_header(struct s2n_stuffer *output, uint8_t record_type, uint8_t message_type)
+{
+    GUARD(s2n_stuffer_write_uint8(output, record_type));
+
+    /* TLS1.2 protocol version */
+    GUARD(s2n_stuffer_write_uint8(output, 3));
+    GUARD(s2n_stuffer_write_uint8(output, 3));
+
+    if (record_type == TLS_HANDSHAKE) {
+        /* Total message size */
+        GUARD(s2n_stuffer_write_uint16(output, 4));
+
+        GUARD(s2n_stuffer_write_uint8(output, message_type));
+
+        /* Handshake message data size */
+        GUARD(s2n_stuffer_write_uint24(output, 0));
+        return 0;
+    }
+
+    if (record_type == TLS_CHANGE_CIPHER_SPEC) {
+        /* Total message size */
+        GUARD(s2n_stuffer_write_uint16(output, 1));
+
+        /* change spec is always just 0x01 */
+        GUARD(s2n_stuffer_write_uint8(output, 1));
+        return 0;
+    }
+
+    return 0;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    /* Construct an array of all valid tls1.3 handshake_types */
+    uint16_t valid_tls13_handshakes[S2N_HANDSHAKES_COUNT];
+    int valid_tls13_handshakes_size = 0;
+    for (int i = 0; i < S2N_HANDSHAKES_COUNT; i++) {
+        if( memcmp(tls13_handshakes, invalid_handshake, S2N_MAX_HANDSHAKE_LENGTH) != 0) {
+            valid_tls13_handshakes[valid_tls13_handshakes_size] = i;
+            valid_tls13_handshakes_size++;
+        }
+    }
+
+    /* Use handler stubs to avoid errors in handler implementation */
+    for (int i = 0; i < sizeof(tls13_state_machine) / sizeof(struct s2n_handshake_action); i++) {
+        tls13_state_machine[i].handler[0] = s2n_test_handler;
+        tls13_state_machine[i].handler[1] = s2n_test_handler;
+    }
+
+    /* Test: When using TLS 1.2, use the existing state machine and handshakes */
+    {
+        struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT);
+        conn->actual_protocol_version = S2N_TLS12;
+        EXPECT_EQUAL(ACTIVE_STATE_MACHINE(conn), state_machine);
+        EXPECT_EQUAL(ACTIVE_HANDSHAKES(conn), handshakes);
+        EXPECT_SUCCESS(s2n_connection_free(conn));
+    }
+
+    /* Test: When using TLS 1.3, use the new state machine and handshakes */
+    {
+        struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT);
+        conn->actual_protocol_version = S2N_TLS13;
+        EXPECT_EQUAL(ACTIVE_STATE_MACHINE(conn), tls13_state_machine);
+        EXPECT_EQUAL(ACTIVE_HANDSHAKES(conn), tls13_handshakes);
+        EXPECT_SUCCESS(s2n_connection_free(conn));
+    }
+
+    /* Test: TLS1.3 server does not wait for client cipher change requests */
+    {
+        struct s2n_connection *conn = s2n_connection_new(S2N_SERVER);
+        conn->actual_protocol_version = S2N_TLS13;
+
+        for (int i=0; i < valid_tls13_handshakes_size; i++) {
```

Comment:
style nit: there's usually space between assignments eg. `i = 0`, similarly for following.